File: frontend/lib/models/order.dart
// Order class - represents a food order
// Tracks what user ordered and its status

class Order {
  String id;
  String restaurantId;
  String restaurantName;
  List<OrderItem> items;
  double totalAmount;
  String status;
  String paymentStatus;
  String createdAt;

  // Constructor
  Order({
    required this.id,
    required this.restaurantId,
    required this.restaurantName,
    required this.items,
    required this.totalAmount,
    required this.status,
    required this.paymentStatus,
    required this.createdAt,
  });

  // Convert JSON to Order object
  factory Order.fromJson(Map<String, dynamic> json) {
    var itemsList = (json['items'] as List?)
            ?.map((item) => OrderItem.fromJson(item))
            .toList() ??
        [];

    // Handle restaurantId - can be either String or populated object
    String restaurantId = '';
    String restaurantName = 'Unknown Restaurant';
    
    if (json['restaurantId'] != null) {
      if (json['restaurantId'] is String) {
        restaurantId = json['restaurantId'];
      } else if (json['restaurantId'] is Map) {
        restaurantId = json['restaurantId']['_id']?.toString() ?? '';
        restaurantName = json['restaurantId']['name'] ?? 'Unknown Restaurant';
      }
    }

    return Order(
      id: json['_id']?.toString() ?? '',
      restaurantId: restaurantId,
      restaurantName: restaurantName,
      items: itemsList,
      totalAmount: (json['totalAmount'] ?? 0).toDouble(),
      status: json['status'] ?? 'PLACED',
      paymentStatus: json['paymentStatus'] ?? 'PENDING',
      createdAt: json['createdAt']?.toString() ?? '',
    );
  }
}

// OrderItem class - individual items in an order
class OrderItem {
  String menuItemId;
  String name;
  double price;
  int quantity;

  OrderItem({
    required this.menuItemId,
    required this.name,
    required this.price,
    required this.quantity,
  });

  factory OrderItem.fromJson(Map<String, dynamic> json) {
    return OrderItem(
      menuItemId: json['menuItemId']?.toString() ?? '',
      name: json['nameSnapshot']?.toString() ?? '',
      price: (json['priceSnapshot'] ?? 0).toDouble(),
      quantity: (json['quantity'] is int) 
          ? json['quantity'] 
          : int.tryParse(json['quantity']?.toString() ?? '1') ?? 1,
    );
  }
}

File: backend/src/models/Order.js
// Order model - represents customer orders
// This is the most important model in our system!

import mongoose from "mongoose";

// Possible order statuses
const ORDER_STATUS = {
  PLACED: "PLACED",       // Customer just placed order
  CONFIRMED: "CONFIRMED", // Restaurant confirmed it
  PREPARING: "PREPARING", // Restaurant is cooking
  READY: "READY",         // Food is ready for pickup
  DELIVERED: "DELIVERED", // Order completed
  CANCELLED: "CANCELLED"  // User or restaurant cancelled
};

// Payment statuses
const PAYMENT_STATUS = {
  PENDING: "PENDING",  // Not paid yet
  PAID: "PAID"         // Payment completed
};

// Define how an Order looks in the database
const orderSchema = new mongoose.Schema(
  {
    // Which customer placed this order
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
      required: true
    },

    // Which restaurant is this order for
    restaurantId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Restaurant",
      required: true
    },

    // List of items in this order
    // We save name and price as "snapshot" because menu can change later
    items: [
      {
        menuItemId: {
          type: mongoose.Schema.Types.ObjectId,
          ref: "MenuItem"
        },
        nameSnapshot: String,   // Save item name at time of order
        priceSnapshot: Number,  // Save item price at time of order
        quantity: {
          type: Number,
          required: true,
          min: 1
        }
      }
    ],

    // Total amount for this order
    totalAmount: {
      type: Number,
      required: true,
      min: 0
    },

    // Current status of the order
    status: {
      type: String,
      enum: Object.values(ORDER_STATUS),
      default: ORDER_STATUS.PLACED
    },

    // Payment status
    paymentStatus: {
      type: String,
      enum: Object.values(PAYMENT_STATUS),
      default: PAYMENT_STATUS.PENDING
    }
  },
  {
    // Automatically add createdAt and updatedAt
    timestamps: true
  }
);

// Create indexes for faster searches
orderSchema.index({ userId: 1 });
orderSchema.index({ restaurantId: 1 });
orderSchema.index({ createdAt: -1 });

// Create the Order model
const Order = mongoose.model("Order", orderSchema);

// Export both the model and status enums
export default Order;
export { ORDER_STATUS, PAYMENT_STATUS };
